File: unit_test/image.cpp
#include "bitmap_image.hpp"
#include <sstream>

using namespace std;

int main(int argc, char* argv[]){
    ifstream file(argv[1]);
    string temp;
    size_t pos;
    int layer = 0;
    int offset_x,offset_y;
    int width,height;
    
    int line_num;
    int bl_x,bl_y;
    int tr_x,tr_y;
    int wire_num;
    int layer_num;
    char normal[7];

    cout<<"Which layer do you want to generate? ";
    cin>>layer;

    if(!file){  // check file exist or not
        cerr<<"Can't open layout file\n";
        exit(-1);
    }
    // Read layout boundary
    getline(file,temp);
    pos = temp.find(";");
    temp = temp.substr(0,pos);
    sscanf(temp.c_str(),"%d %d %d %d",&offset_x,&offset_y,&width,&height);
    width  -= offset_x;
    height -= offset_y;
    cout<<"Offset_x = "<<offset_x<<" Offset_y = "<<offset_y<<endl;
    cout<<"Width = "<<width<<" Height = "<<height<<endl;
    // image constructor
    bitmap_image image(width,height);
    // set background to white
    image.set_all_channels(255,255,255);
    image_drawer draw(image);

    cout<<"image set"<<endl;
    while(getline(file,temp)){
        sscanf(temp.c_str(),"%d %d %d %d %d %d %d %s",
                    &line_num, &bl_x, &bl_y, &tr_x, &tr_y, 
                    &wire_num, &layer_num, normal);
        if(layer != layer_num) break;
        else{
            cout<<bl_x<<" "<<bl_y<<" "<<tr_x<<" "<<tr_y<<endl;
            bl_x -= offset_x;
            tr_x -= offset_x;
            bl_y -= offset_y;
            tr_y -= offset_y;
            cout<<bl_x<<" "<<bl_y<<" "<<tr_x<<" "<<tr_y<<endl;
            draw.pen_width(1);
            draw.pen_color(0,0,255);
            draw.rectangle(bl_x,bl_y,tr_x,tr_y);
        }
    }
    image.save_image("test.bmp");

    return 0;
}

File: unit_test/image.py
##################
#    README
# Plot specific bin
# usage: python3 image.py ./normal.cut ./fill.cut

import sys
import matplotlib.pyplot as plt
from matplotlib.collections import PatchCollection
import matplotlib.pyplot as plt
import matplotlib.patches as patches

if len(sys.argv) < 3:
    print ("Use:", sys.argv[0], "(layout  normal filename) (layout fill filename)")
    sys.exit(1) 
filename = sys.argv[1]
print (filename)


file = open(filename, "r")
line = file.readline() # read boundary
pos = line.find(';')
line = line[:pos]
offset_x, offset_y, width, height = line.split(' ') 
offset_x = int(offset_x)
offset_y = int(offset_y)
width    = int(width) - offset_x
height   = int(height) - offset_y

if(width > height):
    lims = (0,width)
else:
    lims = (0,height)

fig = plt.figure()

for line in iter(file):
    pos = line.find("normal")
    line = line[:pos]
    bl_x, bl_y, tr_x, tr_y, net_num,y = line.split(' ')
    print(bl_x, bl_y, tr_x, tr_y, net_num)
    bl_x = int(bl_x) - offset_x
    bl_y = int(bl_y) - offset_y
    tr_x = int(tr_x) - offset_x - bl_x
    tr_y = int(tr_y) - offset_y - bl_y
    net_num = int(net_num)
    bl_pt = [bl_x,bl_y]
    ax1 = fig.add_subplot(111, aspect='equal')
    if net_num == 0: # not critical
        ax1.add_patch(patches.Rectangle((bl_x,bl_y),tr_x,tr_y, facecolor = "red"))
    else:
        ax1.add_patch(patches.Rectangle((bl_x,bl_y),tr_x,tr_y, facecolor = "blue"))
    
file.close()

filename = sys.argv[2]
file = open(filename, "r")
line = file.readline() # read boundary
pos = line.find(';')
line = line[:pos]
offset_x, offset_y, width, height = line.split(' ') 
offset_x = int(offset_x)
offset_y = int(offset_y)
width    = int(width) - offset_x
height   = int(height) - offset_y


for line in iter(file):
    pos = line.find("fill")
    line = line[:pos]
    bl_x, bl_y, tr_x, tr_y, net_num,y = line.split(' ')
    print(bl_x, bl_y, tr_x, tr_y, net_num)
    bl_x = int(bl_x) - offset_x
    bl_y = int(bl_y) - offset_y
    tr_x = int(tr_x) - offset_x - bl_x
    tr_y = int(tr_y) - offset_y - bl_y
    net_num = int(net_num)
    bl_pt = [bl_x,bl_y]
    ax1 = fig.add_subplot(111, aspect='equal')
    
    ax1.add_patch(patches.Rectangle((bl_x,bl_y),tr_x,tr_y, facecolor = "gray"))
    
file.close()

plt.xlim((0,width))
plt.ylim((0,height))
plt.xlabel('X')
plt.ylabel('Y')

plt.show()